crates/rust-analyzer/src/bin/logger.rs

```diff
@@ -1,88 +1,142 @@
-//! Simple logger that logs either to stderr or to a file, using `env_logger`
-//! filter syntax. Amusingly, there's no crates.io crate that can do this and
-//! only this.
+//! Simple logger that logs either to stderr or to a file, using `tracing_subscriber`
+//! filter syntax and `tracing_appender` for non blocking output.
 
 use std::{
+    fmt::{self, Write},
     fs::File,
-    io::{self, BufWriter, Write},
+    io,
+    sync::Arc,
 };
 
-use env_logger::filter::{Builder, Filter};
-use log::{Log, Metadata, Record};
-use parking_lot::Mutex;
+use rust_analyzer::Result;
+use tracing::{level_filters::LevelFilter, Event, Subscriber};
+use tracing_log::NormalizeEvent;
+use tracing_subscriber::{
+    fmt::{writer::BoxMakeWriter, FmtContext, FormatEvent, FormatFields, FormattedFields},
+    layer::SubscriberExt,
+    registry::LookupSpan,
+    util::SubscriberInitExt,
+    EnvFilter, Registry,
+};
+use tracing_tree::HierarchicalLayer;
 
 pub(crate) struct Logger {
-    filter: Filter,
-    file: Option<Mutex<BufWriter<File>>>,
-    no_buffering: bool,
+    filter: EnvFilter,
+    file: Option<File>,
 }
 
 impl Logger {
-    pub(crate) fn new(log_file: Option<File>, no_buffering: bool, filter: Option<&str>) -> Logger {
-        let filter = {
-            let mut builder = Builder::new();
-            if let Some(filter) = filter {
-                builder.parse(filter);
-            }
-            builder.build()
-        };
+    pub(crate) fn new(file: Option<File>, filter: Option<&str>) -> Logger {
+        let filter = filter.map_or(EnvFilter::default(), |dirs| EnvFilter::new(dirs));
 
-        let file = log_file.map(|it| Mutex::new(BufWriter::new(it)));
-
-        Logger { filter, file, no_buffering }
+        Logger { filter, file }
     }
 
-    pub(crate) fn install(self) {
-        let max_level = self.filter.filter();
-        let _ = log::set_boxed_logger(Box::new(self)).map(|()| log::set_max_level(max_level));
-    }
-}
+    pub(crate) fn install(self) -> Result<()> {
```

Comment:
r? @rust-analyzer/review

This still feels more complicated than it should be, but I'm not opposed to merging it to avoid having a long-running open PR.

Reply:
@lnicola Any progress? :D

Reply:
Sorry for the delay. Can you rebase it?

Reply:
Done.